File: macosx/samples/dec_avc_push/options.cpp
/*
 *  Copyright (c) 2016 Primo Software. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree.  
*/
#include <string.h>
#include <AVBlocks.h>
#include <PrimoReference++.h>
#include <string>
#include <iostream>
#include <iomanip>

#include "options.h"
#include "program_options.h"
#include "util.h"

using namespace std;
using namespace primo::avblocks;	
using namespace primo::codecs;	
using namespace primo::program_options;

void help(OptionsConfig<char>& optionsConfig)
{
	cout << "\nUsage: dec_avc_push --input <directory> [--output <file>]" << endl;
	doHelp(cout, optionsConfig);
}

void setDefaultOptions(Options& opt)
{
    opt.input_dir         = getExeDir() + "/../assets/vid/foreman_qcif.h264.au";
    opt.output_file       = getExeDir() + "/decoded_file.yuv";
}

bool validateOptions(Options& opt)
{  
    return !opt.input_dir.empty() && !opt.output_file.empty();       
}

ErrorCodes prepareOptions(Options& opt, int argc, char* argv[])
{
    if (argc < 2)
    {
        setDefaultOptions(opt);
        cout << "Using defaults:\n";
        cout << " --input "   << opt.input_dir;
        cout << " --output "  << opt.output_file;
        cout << endl;
        return Parsed;
    }

    OptionsConfig<char> optionsConfig;
    optionsConfig.addOptions()
        ("help,?",     opt.help,                           "")
        ("input,i",    opt.input_dir,   string(),          "input directory (contains sequence of compressed files)")
        ("output,o",   opt.output_file, string(),          "output YUV file");

    try
    {
        scanArgv(optionsConfig, argc, argv);
    }
    catch (ParseFailure<char> &ex)
    {
        cout << ex.message() << endl;
        help(optionsConfig);
        return Error;
    }

    if(opt.help)
    {
        help(optionsConfig);
        return Command;
    }

    if (!validateOptions(opt))
    {
        help(optionsConfig);
        return Error;
    }

    return Parsed;
}

File: windows/net/samples/dec_avc_push/Options.cs
/*
 *  Copyright (c) 2016 Primo Software. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree.  
*/
using System;
using System.IO;
using CommandLine;
using CommandLine.Text;
using PrimoSoftware.AVBlocks;

namespace DecAvcPushSample
{
    class Options
    {
        // enc_avc_push command line options
        [Option('?', "help", HelpText = "Display this help screen")]
        public bool Help { get; set; }

        [Option('i', "input", HelpText = "input Access unit(AU) files")]
        public string InputDir { get; set; }

        [Option('o', "output", HelpText = "output YUV file")]
        public string OutputFile { get; set; }

        // The program parses the command line options and sets these properties
        public bool Error { get; private set; }

        string GetUsage()
        {
            return HelpText.AutoBuild(this, (HelpText current) => HelpText.DefaultParsingErrorsHandler(this, current));
        }

        void PrintUsage()
        {
            Console.WriteLine("\nUsage: dec_avc_push --input <AU_Folder> --output <outputFile.yuv>");
            Console.WriteLine(GetUsage());
        }

        void ResetOptions()
        {
            Help = false;
            InputDir = null;
            OutputFile = null;
            Error = false;
        }

        void SetDefaultOptions()
        {
            string exeDir = Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location);
            InputDir = Path.Combine(exeDir, "..\\assets\\vid\\foreman_qcif.h264.au");
            OutputFile = Path.Combine(exeDir, "decoded_file.yuv");

            Console.WriteLine("Using default options: ");
            Console.Write("dec_avc_push --input " + InputDir);
            Console.Write(" --output " + OutputFile);
            Console.WriteLine();
        }

        public bool Prepare(string[] args)
        {
            ResetOptions();

            if (args.Length == 0)
            {
                SetDefaultOptions();
                return true;
            }
            else
            {
                if (!CommandLine.Parser.Default.ParseArguments(args, this))
                {
                    Console.WriteLine("Syntax error");
                    PrintUsage();
                    Error = true;
                    return false;
                }
            }

            if (Help)
            {
                PrintUsage();
                return false;
            }

            if (!Validate())
            {
                Console.WriteLine("\nRequired options are not set!");
                PrintUsage();
                Error = true;
                return false;
            }
            
            return true;
        }

        bool Validate()
        {
            bool res = true;

            if (InputDir == null)
            {
                Console.WriteLine("[not set]");
                res = false;
            }

            if (OutputFile == null)
            {
                Console.WriteLine("[not set]");
                res = false;
            }

            return res;
        }
    }
}
